src/Functions/array/arrayIndex.h

```diff
@@ -574,94 +624,247 @@ struct ArrayIndexGenericNullImpl
+    {
+        return (executeIntegralExpanded<Integral, Integral...>(block, arguments, result) || ...);
+    }
 
-    template <typename T>
-    bool executeNumber(Block & block, const ColumnNumbers & arguments, size_t result)
+    /// Invoke executeIntegralImpl with such parameters: (A, other1), (A, other2), ...
+    template <class A, class ...Other>
+    bool executeIntegralExpanded(Block & block, const ColumnNumbers & arguments, size_t result)
     {
-        return executeNumberNumber<T, UInt8>(block, arguments, result)
-            || executeNumberNumber<T, UInt16>(block, arguments, result)
-            || executeNumberNumber<T, UInt32>(block, arguments, result)
-            || executeNumberNumber<T, UInt64>(block, arguments, result)
-            || executeNumberNumber<T, Int8>(block, arguments, result)
-            || executeNumberNumber<T, Int16>(block, arguments, result)
-            || executeNumberNumber<T, Int32>(block, arguments, result)
-            || executeNumberNumber<T, Int64>(block, arguments, result)
-            || executeNumberNumber<T, Float32>(block, arguments, result)
-            || executeNumberNumber<T, Float64>(block, arguments, result);
+        return (executeIntegralImpl<A, Other>(block, arguments, result) || ...);
     }
 
-    template <typename T, typename U>
-    bool executeNumberNumber(Block & block, const ColumnNumbers & arguments, size_t result)
+    /**
+     * The internal data type of the first argument (target array), if it's integral, like UInt8, may differ from the 
+     * second argument, namely, the @e value, so it's possible to invoke the <tt>has(Array(Int8), UInt64)</tt> e.g. 
+     * so we have to check all possible variants for #Initial and #Resulting types.
+     */
+    template <typename Initial, typename Resulting>
+    bool executeIntegralImpl(Block & block, const ColumnNumbers & arguments, size_t result)
     {
-        const ColumnArray * col_array = checkAndGetColumn<ColumnArray>(block.getByPosition(arguments[0]).column.get());
+        const ColumnArray * col_array = checkAndGetColumn<ColumnArray>(
+                block.getByPosition(arguments[0]).column.get());
 
         if (!col_array)
             return false;
 
-        const ColumnVector<T> * col_nested = checkAndGetColumn<ColumnVector<T>>(&col_array->getData());
+        const ColumnVector<Initial> * col_nested = checkAndGetColumn<ColumnVector<Initial>>(&col_array->getData());
 
         if (!col_nested)
             return false;
 
         auto col_res = ResultColumnType::create();
 
-        /// Null maps of the 1st and second function arguments,
-        /// if it applies.
-        const PaddedPODArray<UInt8> * null_map_data = nullptr;
-        const PaddedPODArray<UInt8> * null_map_item = nullptr;
+        const auto [null_map_data, null_map_item] = nullMapsBuilder(block, arguments);
+        const IColumn* item_arg = block.getByPosition(arguments[1]).column.get();
```

Comment:
also, it is quite irritating to have multiple names for same thing.
e.g. in indexOf(haystack, needle)
* haystack is called: array and data, 
* needle is called: arg, item, rhs, value.

Could you please peek one?

Reply:
It's the original author's decision, not mine.